for-research/public-likelihoods/public-likelihoods.ipynb

```diff
@@ -0,0 +1,936 @@
+    "w2.SaveAs(\"reduced_ws.root\") # BUG: json serializer currently crashing, so can only save as a ROOT file at the moment"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a31df42f-2a30-47f2-ade5-dabc4ecbc3de",
+   "metadata": {},
+   "source": [
+    "# 3. Constructing NLL Functions\n",
+    "\n",
+    "An NLL function (the objective function to be minimized) is constructed from a pdf and a dataset using the `xRooNode::nll` method. \n",
+    "\n",
+    "## 3.1. Choose which parameters to float\n",
+    "\n",
+    "While this isn't strictly necessary to do before constructing the NLL function, there are certain **constant term optimizations** that can be performed if you first mark constant any parameter you plan to hold constant during any fit(s). For this example we will only float the poi and a subset of the nuisance parameters:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "13505cce-887b-4fde-b909-40159a05eb03",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are 12 floating parameters\n"
+     ]
+    }
+   ],
+   "source": [
+    "w.pars().setAttribAll(\"Constant\") # marks all parameters constant\n",
+    "w.poi().setAttribAll(\"Constant\",False) # re-floats all the poi\n",
+    "w.pars().reduced(\"alpha_EG*,gamma_stat_reg_4l_SF*\").setAttribAll(\"Constant\",False) # re-floats some nuisance parameters (usually represent systematics)\n",
+    "print(\"There are\",w.floats().size(),\"floating parameters\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b9381984-8088-420c-9635-8550c6669814",
+   "metadata": {},
+   "source": [
+    "## 3.2. Select NLL Options\n",
+    "\n",
+    "There are a number of NLL-construction options available, [listed here](https://root.cern.ch/doc/master/classRooAbsPdf.html#a24b1afec4fd149e08967eac4285800de). The NLL options can be assembled as follows:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "id": "543d3b7f-973c-4f6a-acaf-6e0550b134f0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "nllOpts = ROOT.xRooFit.createNLLOptions() # use the default xRooFit options as a starting point (enables offsetting, and disables const-term optimization)\n",
+    "nllOpts.Add(ROOT.RooFit.ExternalConstraints(w.pdf(\"regularization_dPhi_ttbar_Z\")).Clone()) # add second pdf as the regularization term for the NLL\n",
+    "nllOpts.Add(ROOT.RooFit.EvalBackend(\"legacy\").Clone()) # use legacy backend evaluation method"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a11f6970-3398-4e55-b930-5ee64083678b",
+   "metadata": {},
+   "source": [
+    "## 3.3. Constructing the NLL with a pdf, dataset, and NLL Options list\n",
+    "\n",
+    "Now you can create the NLL function for your chosen pdf. If you pass \"\" as the dataset name an asimov dataset corresponding to the current state of the model will be used. "
```

Comment:
Again, I think it'd be great if the output could be explained, or if we could point to something where the output (at least terminology) is explained.